File: eve-app/android/app/src/main/java/app/eve/ui/talk/ToolVisuals.kt
package app.eve.ui.talk

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.MenuBook
import androidx.compose.material.icons.automirrored.outlined.ReceiptLong
import androidx.compose.material.icons.automirrored.outlined.Send
import androidx.compose.material.icons.automirrored.outlined.VolumeUp
import androidx.compose.material.icons.outlined.Alarm
import androidx.compose.material.icons.outlined.Bolt
import androidx.compose.material.icons.outlined.Bookmark
import androidx.compose.material.icons.outlined.CalendarMonth
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Computer
import androidx.compose.material.icons.outlined.Edit
import androidx.compose.material.icons.outlined.Email
import androidx.compose.material.icons.outlined.Inbox
import androidx.compose.material.icons.outlined.Insights
import androidx.compose.material.icons.outlined.Newspaper
import androidx.compose.material.icons.outlined.Psychology
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material.icons.outlined.SmartToy
import androidx.compose.material.icons.outlined.WbSunny
import androidx.compose.ui.graphics.vector.ImageVector

/**
 * Per-tool icon + friendly title + present-tense running phrase — the Android port of the desktop
 * `lib/toolVisuals.ts`. The transformation here is text+icon: the UI shows a friendly phrase
 * ("Checking your email…") and a meaningful icon instead of the raw tool id. Unknown tools fall
 * back to a bot icon + a title-cased name + "Running {Name}…" (matches the TS fallback).
 *
 * Note `jarvis_agent` is the tool that triggers the delegation waterfall — its icon is the brain.
 */
data class ToolVisual(val icon: ImageVector, val title: String, val running: String)

fun toolVisual(tool: String): ToolVisual = when (tool) {
    "check_email" -> ToolVisual(Icons.Outlined.Email, "Email", "Checking your email…")
    "check_inbox" -> ToolVisual(Icons.Outlined.Inbox, "Inbox", "Checking your inbox…")
    "get_calendar" -> ToolVisual(Icons.Outlined.CalendarMonth, "Calendar", "Looking at your calendar…")
    "get_news" -> ToolVisual(Icons.Outlined.Newspaper, "News", "Pulling the headlines…")
    "get_weather" -> ToolVisual(Icons.Outlined.WbSunny, "Weather", "Checking the weather…")
    "open_on_pc" -> ToolVisual(Icons.Outlined.Computer, "Open on PC", "Opening that on your PC…")
    "search_knowledge" -> ToolVisual(Icons.Outlined.Search, "Knowledge", "Searching your knowledge…")
    "search_notes" -> ToolVisual(Icons.Outlined.Search, "Notes", "Searching your notes…")
    "send_to_channel" -> ToolVisual(Icons.AutoMirrored.Outlined.Send, "Message", "Sending your message…")
    "set_reminder" -> ToolVisual(Icons.Outlined.Alarm, "Reminder", "Setting your reminder…")
    "set_voice" -> ToolVisual(Icons.AutoMirrored.Outlined.VolumeUp, "Voice", "Switching my voice…")
    "start_challenger_mode" -> ToolVisual(Icons.Outlined.Bolt, "Challenger", "Entering challenger mode…")
    "system_report" -> ToolVisual(Icons.Outlined.Insights, "System report", "Auditing my own tools…")
    "remember" -> ToolVisual(Icons.Outlined.Bookmark, "Remember", "Saving that to memory…")
    "recall" -> ToolVisual(Icons.AutoMirrored.Outlined.MenuBook, "Recall", "Recalling what I know…")
    "prepare_text" -> ToolVisual(Icons.Outlined.Edit, "Draft text", "Drafting the text…")
    "confirm_send_text" -> ToolVisual(Icons.AutoMirrored.Outlined.Send, "Send text", "Sending the text…")
    "jarvis_agent" -> ToolVisual(Icons.Outlined.Psychology, "Agent", "Handing this to the agent…")
    "create_invoice" -> ToolVisual(Icons.AutoMirrored.Outlined.ReceiptLong, "Invoice", "Drafting the invoice…")
    "confirm_action" -> ToolVisual(Icons.Outlined.CheckCircle, "Confirm", "Confirming…")
    else -> {
        val name = titleCaseTool(tool)
        ToolVisual(Icons.Outlined.SmartToy, name, "Running $name…")
    }
}

/** "open_on_pc" → "Open On Pc". Mirrors the TS fallback's title-casing of an unknown tool id. */
private fun titleCaseTool(tool: String): String =
    tool.split('_', ' ', '-')
        .filter { it.isNotBlank() }
        .joinToString(" ") { part -> part.replaceFirstChar { it.uppercaseChar() } }
        .ifBlank { "Tool" }

File: app/frontend/src/lib/toolVisuals.ts
// Maps an EVE tool name to a friendly, human-readable visual: an icon + a present-tense
// "running" phrase ("Checking your email…") + a short noun title ("Email"). This is what makes
// EVE's talking UI *do something* when it invokes a tool, instead of showing the raw tool id.
//
// Pure + framework-agnostic (returns a lucide icon component reference, no JSX) so it is unit
// tested in isolation. Unknown tools fall back to a wrench + a prettified name.
import {
  Activity,
  AlarmClock,
  BookOpen,
  Bookmark,
  Bot,
  Brain,
  Calendar,
  Camera,
  CheckCircle2,
  CloudSun,
  Inbox,
  Mail,
  MessageSquare,
  Monitor,
  MonitorUp,
  Newspaper,
  Receipt,
  Search,
  Send,
  Swords,
  Volume2,
  type LucideIcon,
} from 'lucide-react';

export interface ToolVisual {
  /** Lucide icon representing the tool. */
  Icon: LucideIcon;
  /** Short noun shown once the tool finishes, e.g. "Email", "Invoice". */
  title: string;
  /** Present-tense status shown while running, e.g. "Checking your email…". */
  running: string;
}

const TABLE: Record<string, ToolVisual> = {
  check_email: { Icon: Mail, title: 'Email', running: 'Checking your email…' },
  check_inbox: { Icon: Inbox, title: 'Inbox', running: 'Checking your inbox…' },
  get_calendar: { Icon: Calendar, title: 'Calendar', running: 'Looking at your calendar…' },
  get_news: { Icon: Newspaper, title: 'News', running: 'Pulling the headlines…' },
  get_weather: { Icon: CloudSun, title: 'Weather', running: 'Checking the weather…' },
  open_on_pc: { Icon: Monitor, title: 'Open on PC', running: 'Opening that on your PC…' },
  search_knowledge: { Icon: Search, title: 'Knowledge', running: 'Searching your knowledge base…' },
  search_notes: { Icon: Search, title: 'Notes', running: 'Searching your notes…' },
  send_to_channel: { Icon: Send, title: 'Message', running: 'Sending your message…' },
  set_reminder: { Icon: AlarmClock, title: 'Reminder', running: 'Setting your reminder…' },
  set_voice: { Icon: Volume2, title: 'Voice', running: 'Switching my voice…' },
  start_challenger_mode: { Icon: Swords, title: 'Challenger', running: 'Entering challenger mode…' },
  system_report: { Icon: Activity, title: 'System report', running: 'Auditing my own tools…' },
  remember: { Icon: Bookmark, title: 'Remember', running: 'Saving that to memory…' },
  recall: { Icon: BookOpen, title: 'Recall', running: 'Recalling what I know…' },
  prepare_text: { Icon: MessageSquare, title: 'Draft text', running: 'Drafting a text…' },
  confirm_send_text: { Icon: Send, title: 'Send text', running: 'Sending the text…' },
  jarvis_agent: { Icon: Brain, title: 'Agent', running: 'Handing this to the agent…' },
  create_invoice: { Icon: Receipt, title: 'Invoice', running: 'Drafting the invoice…' },
  confirm_action: { Icon: CheckCircle2, title: 'Confirm', running: 'Confirming…' },
  look_via_phone: { Icon: Camera, title: 'Looking', running: 'Looking through your phone camera…' },
  surface_visual: { Icon: MonitorUp, title: 'On screen', running: 'Putting that on your screen…' },
};

/** Title-case a raw tool id: "get_weather" → "Get Weather". */
export function prettifyToolName(tool: string): string {
  return tool
    .split(/[_\s]+/)
    .filter(Boolean)
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ');
}

/** Resolve a tool name to its visual, falling back gracefully for unknown tools. */
export function toolVisual(tool: string): ToolVisual {
  const known = TABLE[tool?.toLowerCase?.() ?? ''];
  if (known) return known;
  const pretty = prettifyToolName(tool || 'tool');
  return { Icon: Bot, title: pretty, running: `Running ${pretty}…` };
}
